backtracking/remove_invalid_parenthese.py

```diff
@@ -0,0 +1,120 @@
+def remove_invalid_parentheses(input_string: str) -> list[str]:
+    """
+    Removes the minimum number of invalid parentheses
+    to make the input string valid.
+
+    Args:
+        input_string (str): The input string containing parentheses.
+
+    Returns:
+        List[str]: A list of valid strings after removing
+        the minimum number of parentheses.
+
+    Examples:
+        >>> remove_invalid_parentheses(")(a)())")
+        ['(a())', '(a)()']
+
+        >>> remove_invalid_parentheses("()())()")
+        ['(())()', '()()()']
+
+        >>> remove_invalid_parentheses(")(")
+        ['']
+    """
+
+    def is_valid(next_string: str) -> bool:
+        """
+        Helper function to check if a string has valid parentheses.
+
+        Params:
+            next_string (str): The input string to be checked.
+        Returns:
+            bool: True if the parentheses in the string are valid, False otherwise.
+
+        Examples:
+        >>> is_valid("(a)())")
+        False
+
+        >>> is_valid("(a())")
+        True
+
+        >>> is_valid(")(a)()")
+        False
+
+        """
+        count: int = 0
+        for char in next_string:
+            if char == "(":
+                count += 1
+            elif char == ")":
+                count -= 1
+                if count < 0:
+                    return False
+        return count == 0
+
+    def dfs(
+        current_string: str, start_index: int, left_removed: int, right_removed: int
+    ) -> None:
```

Comment:
What is the value of a search that returns `None`?!?

Reply:
Its main objective is not to generate a single result to be returned, but rather to directly edit the valid_parentheses set.

Reply:
Then return the "parentheses set" so that tests can be written.

Reply:
the function named remove_invalid_parentheses returns that set
and
the dfs(depth_first_search) function only searchs that parentheses using backtracking not returning anything